### ISSUE TITLE
Profiling tool for DDP and FSDP

### ISSUE THREAD

**Issue by whatdhack** [pytorch/pytorch#121181]
### 🚀 The feature, motivation and pitch

A good profiling tool appears to be lacking for both DDP and FSDP.

### Alternatives

None. 

### Additional context

Something like Horovod Timeline but better ?

**Comment by awgu** [comment]
What info would you like to see in the profile?

**Comment by whatdhack** [comment]
At the least everything that is there in Horovod Timeline -  refer to this https://horovod.readthedocs.io/en/latest/timeline_include.html  .

**Comment by whatdhack** [comment]
Yes, for single GPU profiling .  Do I get information on DDP and FSDP with the torch_profiler (e.g.  NCCL/MPI  allreduce/allgather  etc ) ?  How do I use it with DDP and FSDP ?  Any documentation will be helpful.

**Comment by yf225** [comment]
@whatdhack you can see NCCL/MPI allreduce/allgather in the torch profiler GPU trace, including input shapes and kernel runtime etc

**Comment by whatdhack** [comment]
Thanks for pointing that out. Abel to see the c10d and nccl calls in the traces.  Quick follow on questions. 

1. mpi4py  appears to come in conjunction with DDP/FSDP - is it really necessary for them to work correctly ?  To me it looks like mpi4py is basically  used for rank and size info. 
2.  DDP  has has an argument  called  bucket_cap_mb.   Looks like  the minimum number of buckets is 2.  Gradients are always spilt between the 2 buckets even when the   aggregate size of gradients  is 1/10 th size of the bucket. 
3.  From the traces it looks like DDP uses a smattering of allgather before starting the allreduce's, Any documentation  on these aspects ?

**Comment by awgu** [comment]
For 3, if you use `with_stack=True` in the profiler, you will be able to see the Python stack trace and where the collective calls are coming from.

**Comment by whatdhack** [comment]
Any update ?

**Comment by awgu** [comment]
For 3, I think it is from here:
https://github.com/pytorch/pytorch/blob/2f7cfecd86009a9d396fdbdcdfb4ba7a005db16b/torch/csrc/distributed/c10d/reducer.cpp#L2237-L2247

**Comment by awgu** [comment]
For 2, https://github.com/pytorch/pytorch/pull/121640 would allow you to just have 1 bucket I think.

**Comment by awgu** [comment]
For 1, I am not sure. Is mpi4py specific to MPI backend for the process group?

**Comment by whatdhack** [comment]
No  1, mpi4py is not needed.  Any  launcher ( e.g. mpirun, etc.) can provide the necessary info.   
No  2., the PR you are pointing to would certainly help,  but  it was no clear if it can  create only one bucket. 
No  3., allgather is used for "DDP expects same model across all ranks, but Rank ".   Is it really necessary ?  Seems  redundant to me.

**Comment by awgu** [comment]
> No 3., allgather is used for "DDP expects same model across all ranks, but Rank ". Is it really necessary ? Seems redundant to me.

I think I was wrong. That all-gather should only be run at construction time. I am not sure about any all-gathers during forward/backward. (Searching for `allgather` in `reducer.cpp` shows nothing, and searching for `all_gather` or `allgather` in `nn/parallel/distributed.py` shows nothing.)
